pkg/controller/shared_state.go

```diff
@@ -1345,13 +1371,160 @@ func (c *SharedState) createReplicas(ctx context.Context, remainingReplicas int,
 		//no failed replica attachments, but there are still more replicas to reach MaxShares
 		request := ReplicaRequest{VolumeName: volumeName, Priority: remainingReplicas}
 		c.priorityReplicaRequestsQueue.Push(ctx, &request)
-		for _, pod := range pods {
-			c.eventRecorder.Eventf(pod.DeepCopyObject(), v1.EventTypeWarning, consts.ReplicaAttachmentFailedEvent, "Not enough suitable nodes to attach %d of %d replica mount(s) for volume %s", remainingReplicas, requiredReplicas, volumeName)
+		message := fmt.Sprintf("Not enough suitable nodes to attach %d of %d replica mount(s) for volume %s", remainingReplicas, requiredReplicas, volumeName)
+		podCopies := make([]runtime.Object, len(pods))
+		for i, pod := range pods {
+			podCopies[i] = pod.DeepCopyObject()
 		}
+		timestamp := time.Now()
+		for _, podCopy := range podCopies {
+			c.eventRecorder.Eventf(podCopy, v1.EventTypeWarning, consts.ReplicaAttachmentFailedEvent, message)
+		}
+		message += timestamp.UTC().Format(" [0102 15:04:05") // append original timestamp
+		c.persistAttachmentFailure(volumeName, message, podCopies, timestamp)
 	}
 	return nil
 }
 
+func (c *SharedState) persistAttachmentFailure(volumeName string, message string, podCopies []runtime.Object, timestamp time.Time) {
+	c._eventRefresherInitOnce.Do(c._eventRefresherInitFunc)
+	c.eventsToPersistQueue <- ReplicaAttachmentFailureInfo{volumeName, message, podCopies, timestamp}
+}
+
+// Only requests an event un-persist if the event refresher is active
+func (c *SharedState) unpersistAttachmentFailure(volumeName string) {
+	trySend(c.eventsToUnpersistQueue, volumeName) // also mitigates unlikely race condition of sending to closed channel
+}
+
+func (c *SharedState) _eventRefresherInitFunc() {
+	c.eventsToPersistQueue = make(chan ReplicaAttachmentFailureInfo, maxPersistEventConcurrency)
+	c.eventsToUnpersistQueue = make(chan string, maxUnpersistEventConcurrency)
+	go c._eventRefresherRoutine()
+}
+
+func (c *SharedState) _eventRefresherRoutine() {
+	var eventTTL = time.Duration(c.config.EventTTLInSec) * time.Second
+	eventMap := map[string]*eventRefresherNode{}
+	newFailureInfo := <-c.eventsToPersistQueue
+	klog.V(2).Infof("Starting the event refresher") // should this be logged via workflow instead?
```

Comment:
Logging the start of the event refresher only after the first persist request comes doesn't really have much value. It would result in a clear flow if we simply entered the select wait loop and processed the persist request there. It's a little confusing as-is because the receiving a new persist request appears to do nothing by breaking out of the inner for loop. It's only if you follow the loop until its end that you realize it loops back to the outer for loop and process the request there.

Reply:
Logging the start of the event refresher is just an additive log message to let us know when the event refresher is first started, w/ messages for when it's deactivated, and reactivated, to help us know how its operating, like if we ever need to debug a potential decrease in overall performance b/c of it.

There are currently 3 entry-points to a new failure for the event refresher (when it's started for the first time, when receives a new event while active, and when it receives a new event while inactive), so we can't handle new failures by simply moving the new failure logic into that waiting for a new event case, an alternative could be a NewFailure section w/ goto's... thoughts? or other recommendations?

Reply:
Given that this is a single goroutine processing the list, it isn't very likely to directly impact performance. The most likely impact would be caused by the additional PUT calls to API server leading to client-side throttling. Since event creation is actually asynchronous (it spawns off a goroutine to do the PUT), this impact may not be directly correlated to these debug traces but we can rely on the API client throttling traces for detection. However, the debug traces *do* impact logging because the driver is typically run with verbosity of 3 in production. While the debug traces are useful to debug operation, they shouldn't be enabled in prod. I would suggest using level 6 for them.

Do we really need to have an "inactive" state for the refresher? It's seems to be this is to support unconditional "un-persist" of the event but we do log a success event when the replica is attached. This could be used to trigger the un-persist and enable the refresher to be "active" always. This would also allow you to simplify the loop.

Reply:
sending to the event persistence queue / channel still includes a slight window of locking & channel operations etc on both the producer & consumer side which we could avoid altogether during attaches when we know there are no events currently being persisted.

we always attempt un-persisting an attachment failure, b/c errors like the volume no longer existing might make that attachment failure no longer relevant & if we retry attaching a volume and it fails again IMHO we should stop refreshing the old failure events for that volume & start refreshing the new ones